zhaquirks/tuya/ts0601_energy_meter.py

```diff
@@ -0,0 +1,1261 @@
+        TuyaZBElectricalMeasurement.AttributeDefs.reactive_power_ph_b.name,
+        TuyaZBElectricalMeasurement.AttributeDefs.reactive_power_ph_c.name,
+        TuyaZBElectricalMeasurement.AttributeDefs.rms_current.name,
+        TuyaZBElectricalMeasurement.AttributeDefs.rms_current_ph_b.name,
+        TuyaZBElectricalMeasurement.AttributeDefs.rms_current_ph_c.name,
+    )
+
+    def update_attribute(self, attr_name: str, value):
+        """Update the cluster attribute."""
+        value = self.energy_direction_mitigation_handler(attr_name, value)
+        attr_name, value = self.energy_direction_handler(attr_name, value)
+        super().update_attribute(attr_name, value)
+        self._update_measurement_type(attr_name)
+        self.virtual_channel_handler(attr_name)
+
+    def _update_measurement_type(self, attr_name: str):
+        """Derive the measurement_type from reported attributes."""
+        if attr_name not in self._ATTRIBUTE_MEASUREMENT_TYPES:
+            return
+        measurement_type = 0
+        for measurement, mask in self._ATTRIBUTE_MEASUREMENT_TYPES.items():
+            if measurement == attr_name or self.get(measurement) is not None:
+                measurement_type |= mask
+        super().update_attribute(
+            self.AttributeDefs.measurement_type.name, measurement_type
+        )
+
+
+class TuyaMetering(
+    VirtualChannelHelper,
+    EnergyDirectionMitigationHelper,
+    EnergyDirectionHelper,
+    MeterClusterHelper,
+    TuyaLocalCluster,
+    TuyaZBMeteringClusterWithUnit,
+):
+    """Metering cluster for Tuya energy meter devices."""
+
+    WATT_MULTIPLIER = 10
+    WATT_HOUR_MULTIPLIER = 1000
+    DECIWATT_HOUR_MULTIPLIER = 100
+
+    @staticmethod
+    def format(
+        whole_digits: int, dec_digits: int, suppress_leading_zeros: bool = True
+    ) -> int:
+        """Return the formatter value for summation and demand Metering attributes."""
+        assert 0 <= whole_digits <= 7, "must be within range of 0 to 7."
+        assert 0 <= dec_digits <= 7, "must be within range of 0 to 7."
+        return (suppress_leading_zeros << 6) | (whole_digits << 3) | dec_digits
+
+    _CONSTANT_ATTRIBUTES: dict[int, Any] = {
+        **TuyaZBMeteringClusterWithUnit._CONSTANT_ATTRIBUTES,
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.status.id: 0x00,
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.multiplier.id: 1,
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.divisor.id: 10000,  # 2 decimals for summation attributes
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.summation_formatting.id: format(
+            whole_digits=7, dec_digits=2
+        ),
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.demand_formatting.id: format(
+            whole_digits=7, dec_digits=1
+        ),
+    }
+
+    _EXTENSIVE_ATTRIBUTES: tuple[str] = (
+        TuyaZBMeteringClusterWithUnit.AttributeDefs.instantaneous_demand.name,
+    )
```

Comment:
I think this should be the unsigned attribute:
```suggestion
    _EXTENSIVE_ATTRIBUTES: tuple[str] = (
        TuyaZBMeteringClusterWithUnit.AttributeDefs.instantaneous_demand.name
        + EnergyDirectionHelper.UNSIGNED_ATTR_SUFFIX,
    )
```